File: tests/test_department_events.js
// Tests for lib/department_events.js (EOS Phase 2, Round 2, 2026-07-19):
// the JS-side shared JSONL correlation-index envelope. Same convention
// as tests/test_recovery_log.js/test_n8n_notify.js.
//
//   node --test tests/test_department_events.js

const test = require('node:test');
const assert = require('node:assert/strict');
const fs = require('fs');
const os = require('os');
const path = require('path');

const { emit, readEvents, VALID_DEPARTMENTS } = require('../lib/department_events.js');

function tmpLogPath() {
  return path.join(fs.mkdtempSync(path.join(os.tmpdir(), 'dept-events-test-')), 'department_events.jsonl');
}

test('emit: unknown department throws', () => {
  const logPath = tmpLogPath();
  assert.throws(() => emit({ department: 'not_a_real_department', event_type: 'x' }, logPath));
});

test('emit: real department writes a real envelope', () => {
  const logPath = tmpLogPath();
  const record = emit({ department: 'golden_hunter', event_type: 'opportunity.attempted', ref_id: 'niche-x', source_log: 'data/golden_hunter_events.jsonl', summary: 'test' }, logPath);
  assert.equal(record.department, 'golden_hunter');
  assert.ok(record.event_id);
  assert.ok(record.emitted_at);
  const entries = readEvents(logPath);
  assert.equal(entries.length, 1);
  assert.equal(entries[0].ref_id, 'niche-x');
});

test('emit: never duplicates business data, envelope-only', () => {
  const logPath = tmpLogPath();
  const record = emit({ department: 'ai_capability_manager', event_type: 'capability.request_logged', ref_id: 'req-1' }, logPath);
  assert.deepEqual(
    Object.keys(record).sort(),
    ['department', 'emitted_at', 'event_id', 'event_type', 'ref_id', 'source_log', 'summary'].sort(),
  );
});

test('emit: append-only, never overwrites', () => {
  const logPath = tmpLogPath();
  emit({ department: 'recovery', event_type: 'recovery.action_taken' }, logPath);
  emit({ department: 'recovery', event_type: 'recovery.action_taken' }, logPath);
  const entries = readEvents(logPath);
  assert.equal(entries.length, 2);
});

test('VALID_DEPARTMENTS: includes all 12 named departments', () => {
  assert.equal(VALID_DEPARTMENTS.length, 12);
  for (const d of ['researchers', 'customer_intelligence', 'golden_hunter', 'pioneer']) {
    assert.ok(VALID_DEPARTMENTS.includes(d));
  }
});

test('readEvents: missing file reads as empty, never throws', () => {
  const entries = readEvents('/no/such/department_events.jsonl');
  assert.deepEqual(entries, []);
});

File: tests/test_department_events.py
"""Tests for department_events.py (EOS Phase 2, Round 2, 2026-07-19):
the Python-side shared JSONL correlation-index envelope.

Runs with stdlib unittest. Every function reads/writes only temp-file-
isolated paths -- never the real data/department_events.jsonl.

    python -m unittest tests.test_department_events -v
"""

import os
import sys
import tempfile
import unittest
from pathlib import Path

_FACTORY_ROOT = Path(__file__).resolve().parent.parent
if str(_FACTORY_ROOT) not in sys.path:
    sys.path.insert(0, str(_FACTORY_ROOT))

import department_events as de


def _temp_path():
    fd, path = tempfile.mkstemp(suffix=".jsonl")
    os.close(fd)
    os.remove(path)
    return path


class TestEmit(unittest.TestCase):
    def setUp(self):
        self.path = _temp_path()

    def tearDown(self):
        if os.path.exists(self.path):
            os.remove(self.path)

    def test_unknown_department_raises(self):
        with self.assertRaises(ValueError):
            de.emit("not_a_real_department", "some.event", path=self.path)

    def test_real_department_writes_a_real_envelope(self):
        record = de.emit("golden_hunter", "opportunity.attempted", ref_id="niche-x",
                          source_log="data/golden_hunter_events.jsonl", summary="test", path=self.path)
        self.assertEqual(record["department"], "golden_hunter")
        self.assertIn("event_id", record)
        self.assertIn("emitted_at", record)
        entries = de.read_events(path=self.path)
        self.assertEqual(len(entries), 1)
        self.assertEqual(entries[0]["ref_id"], "niche-x")

    def test_never_duplicates_business_data_envelope_only(self):
        record = de.emit("ai_capability_manager", "capability.request_logged", ref_id="req-1", path=self.path)
        # Envelope-only guarantee: no arbitrary business fields beyond the schema.
        self.assertEqual(set(record.keys()), {"event_id", "emitted_at", "department", "event_type", "ref_id", "source_log", "summary"})

    def test_append_only_never_overwrites(self):
        de.emit("recovery", "recovery.action_taken", path=self.path)
        de.emit("recovery", "recovery.action_taken", path=self.path)
        entries = de.read_events(path=self.path)
        self.assertEqual(len(entries), 2)

    def test_write_failure_degrades_honestly_never_throws(self):
        record = de.emit("golden_hunter", "x", path="\x00/bad/path.jsonl")
        self.assertTrue(record.get("_log_failed"))


class TestReadEvents(unittest.TestCase):
    def test_missing_file_reads_as_empty_never_throws(self):
        entries = de.read_events(path="/no/such/department_events.jsonl")
        self.assertEqual(entries, [])


if __name__ == "__main__":
    unittest.main()
